x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java

```diff
@@ -506,7 +586,61 @@ public Expression surrogate(SearchStats searchStats) {
             field(),
             buckets(),
             searchStats,
-            (interval, minValue, maxValue) -> getDateRounding(FoldContext.small(), minValue, maxValue)
+            (interval, minValue, maxValue) -> getDateRounding(field, buckets, from, to, FoldContext.small(), minValue, maxValue)
         );
     }
+
+    record DatetimeEmptyBucketGenerator(long from, long to, Rounding.Prepared rounding) implements BlockHash.EmptyBucketGenerator {
+
+        DatetimeEmptyBucketGenerator(Expression field, Expression buckets, Expression from, Expression to, FoldContext foldContext) {
+            this(foldToLong(foldContext, from), foldToLong(foldContext, to), getDateRounding(field, buckets, from, to, foldContext));
+        }
+
+        @Override
+        public int getEmptyBucketCount() {
+            int i = 0;
+            for (long bucket = rounding.round(from); bucket < to; bucket = rounding.nextRoundingValue(bucket)) {
+                i++;
+            }
+            return i;
+        }
+
+        @Override
+        public void generate(Block.Builder blockBuilder) {
+            for (long bucket = rounding.round(from); bucket < to; bucket = rounding.nextRoundingValue(bucket)) {
+                ((LongBlock.Builder) blockBuilder).appendLong(bucket);
+            }
+        }
+    }
+
+    record NumericEmptyBucketGenerator(double from, double to, double roundTo) implements BlockHash.EmptyBucketGenerator {
+
+        NumericEmptyBucketGenerator(Expression buckets, Expression from, Expression to, FoldContext foldContext) {
+            this(
+                ((Number) from.fold(foldContext)).doubleValue(),
+                ((Number) to.fold(foldContext)).doubleValue(),
+                determineRounding(buckets, from, to, foldContext)
+            );
+        }
+
+        @Override
+        public int getEmptyBucketCount() {
+            int i = 0;
+            for (double bucket = round(Math.floor(from / roundTo) * roundTo, 2); bucket < to; bucket = round(bucket + roundTo, 2)) {
+                i++;
+            }
+            return i;
+        }
+
+        @Override
+        public void generate(Block.Builder blockBuilder) {
+            for (double bucket = round(Math.floor(from / roundTo) * roundTo, 2); bucket < to; bucket = round(bucket + roundTo, 2)) {
+                ((DoubleBlock.Builder) blockBuilder).appendDouble(bucket);
```

Comment:
Let's move this cast outside the for. Maybe an:
```
if (blockBuilder instanceof DoubleBlock.Builder doubleBlockBuilder) {
    // Code...
} else {
    throw ...;
}
```

Same for the other generator.